### PR TITLE
Update opcodes

### LINKED ISSUE #209: Sync new opcode: ASSERT,  ABORT
waiting for https://github.com/neo-project/neo-vm/pull/300

### PR DESCRIPTION
**Pull request by shargon**
Close #209

- Update Assert and Abort tests
- Create UT for HexToBytes

### CODE REVIEW

File: src/Neo.SmartContract.Framework/SmartContract.cs [hunk 1]
```diff
@@ -2,6 +2,10 @@ namespace Neo.SmartContract.Framework
 {
     public class SmartContract
     {
-
+        ///// <summary>
+        ///// Faults if b is false
+        ///// </summary>
+        //[OpCode(OpCode.ASSERT)]
+        //public extern static void Assert(bool b);
```
Comment (shargon) [inline src/Neo.SmartContract.Framework/SmartContract.cs]: @lightszero could you check why I can create this call?

File: tests/Neo.SmartContract.Framework.UnitTests/TestClasses/Contract_Helper.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            (value == true).Assert();
+            return 5;
+        }
+
+        public static void VoidAssertExtension(bool value)
+        {
+            (value == true).Assert();
+        }
+
+        //public static void AssertCall(bool value)
```
Comment (shargon) [inline tests/Neo.SmartContract.Framework.UnitTests/TestClasses/Contract_Helper.cs]: I want to allow both ways

**Comment by Tommo-L** [comment]
close #209

File: src/Neo.SmartContract.Framework/Helper.cs [hunk 3]
```diff
@@ -74,10 +74,16 @@ public static class Helper
         public extern static bool Within(this int x, BigInteger a, BigInteger b);
 
         /// <summary>
-        /// Faults iff b is false
+        /// Faults if b is false
         /// </summary>
-        [OpCode(OpCode.THROWIFNOT)]
-        public extern static void Assert(this bool b);
+        [OpCode(OpCode.ASSERT)]
+        public extern static void Assert(this bool condition);
```
Comment (erikzhang) [inline src/Neo.SmartContract.Framework/Helper.cs]: I don't think this is suitable for implementation as an extension method.

File: src/Neo.SmartContract.Framework/SmartContract.cs (hunk 1 above)
Comment (erikzhang) [reply src/Neo.SmartContract.Framework/SmartContract.cs]: We should move `Assert()` and `Abort()` to here.

Comment (shargon) [reply src/Neo.SmartContract.Framework/SmartContract.cs]: Agree, @lightszero could you check it why I can't define here this method?

File: src/Neo.SmartContract.Framework/Helper.cs (hunk 3 above)
Comment (shargon) [reply src/Neo.SmartContract.Framework/Helper.cs]: Ok! I will remove it when the other works

**Comment by shargon** [comment]
Now got this error @lightszero @erikzhang 

`Initialization method Neo.SmartContract.Framework.UnitTests.Services.Neo.IteratorTest.Init threw exception. System.TypeLoadException: Could not load type 'Neo.SmartContract.Framework.SmartContract' from assembly 'Neo.SmartContract.Framework, Version=3.0.0.0, Culture=neutral, PublicKeyToken=null' because the method 'Assert' has no implementation (no RVA)..`

**Comment by shargon** [comment]
I can fix it removing the `abstract` keyword, but i am not sure if it will add a empty CALL with empty body after the call

https://github.com/neo-project/neo-devpack-dotnet/pull/216/commits/d5982079dba4413f99e61123285b3f3c1c0e3c8b

**Comment by lightszero** [comment]
i will check it later.

**Comment by lightszero** [comment]
rosyln build  in unittest can't extend a class with extern method.
@shargon 
so we move Abort and Assert to another class will fix this.

File: src/Neo.SmartContract.Framework/Debug.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Neo.SmartContract.Framework
+{
+    public static class Debug
```
Comment (shargon) [inline src/Neo.SmartContract.Framework/Debug.cs]: Runtime?

**Comment by shargon** [comment]
@lightszero with empty body it's not possible to do that? discard the body if this method has `OpCodeAttribute`?

File: src/Neo.SmartContract.Framework/Debug.cs (hunk 4 above)
Comment (lightszero) [reply src/Neo.SmartContract.Framework/Debug.cs]: you can rename it.

**Comment by lightszero** [comment]
> @lightszero with empty body it's not possible to do that? discard the body if this method has `OpCodeAttribute`?

is possible,but we use extern in framework everywhere.

**Comment by shargon** [comment]
nice!!!

**Comment by erikzhang** [comment]
Test failed.

**Comment by shargon** [comment]
I will revert the last change because it will require `GasLeft`

**Comment by ProDog** [comment]
I have tested it passed.
Contract:

image

Invoke result:
for assert(true):

image
image

for assert(false):

image

for abort:

image

**Comment by ProDog** [comment]
plz review it and merge code.

**Comment by ProDog** [comment]
Can merge into master? @erikzhang @lightszero

**Comment by lightszero** [comment]
> Can merge into master? @erikzhang @lightszero

good for me

**Pull request merged**